### PR TITLE
feat: support access log formatters

### LINKED ISSUE #3279: Support access log formatters in EnvoyProxy
*Description*:
> EG supports access log [^1], no formatters.
>
> Envoy support formatters [^2] when setting access log, such as `envoy.formatter.req_without_query` [^3], it will remove the query string and is used to avoid logging any sensitive information into the access log. It's very important for production usage.

[optional *Relevant Links*:]
[^1]: https://gateway.envoyproxy.io/latest/api/extension_types/#proxyaccesslogformat
[^2]: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/substitution_format_string.proto#envoy-v3-api-msg-config-core-v3-substitutionformatstring
[^3]: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/formatter/req_without_query/v3/req_without_query.proto#extension-envoy-formatter-req-without-query

### PR DESCRIPTION
**Pull request by aoledk**
**What this PR does / why we need it**:

Enable user to configure access log formatter [^1].

**Which issue(s) this PR fixes**:

Fixes #3279

[^1]: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/substitution_format_string.proto#envoy-v3-api-field-config-core-v3-substitutionformatstring-formatters

### CODE REVIEW

File: api/v1alpha1/accesslogging_types.go [hunk 1]
```diff
@@ -55,8 +55,33 @@ type ProxyAccessLogFormat struct {
...
 	JSON map[string]string `json:"json,omitempty"`
+	// Formatters defines a collection of [accesslog formatter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/substitution_format_string.proto#envoy-v3-api-field-config-core-v3-substitutionformatstring-formatters)
+	// that enable additional [command operators](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#command-operators)
+	// that can be used in the accesslog format when accesslog sinks only contain "File".
+	// Valid AccessLogFormatterType values are
+	// "Cel"
+	// "Metadata"
+	// "ReqWithoutQuery"
+	// +optional
+	Formatters []AccessLogFormatterType `json:"formatters,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/accesslogging_types.go:66]: didn't need an api for this, check https://github.com/istio/istio/blob/0ee031b7fe22d23fce0f75a4fb0c38be35a07215/pilot/pkg/model/telemetry_logging.go#L317

Comment (aoledk) [reply api/v1alpha1/accesslogging_types.go:66]: That's a good idea to add formatters based on whether specified operator command shows in text automatically. I'll reimplement this feature in the same way.

File: internal/xds/translator/testdata/in/xds-ir/accesslog-formatters.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+name: "accesslog"
+accesslog:
+  text:
+  - path: "/dev/stdout"
+    format: |
+      [%START_TIME%] "%CEL(request.method)% %CEL(request.path)% %CEL(request.protocol)%" %CEL(response.code)% %CEL(response.flags)% %CEL(request.total_size)% %CEL(response.total_size)% %CEL(request.duration)% %CEL(response.headers['X-ENVOY-UPSTREAM-SERVICE-TIME'])% "%CEL(request.headers['X-FORWARDED-FOR'])%" "%CEL(request.useragent)%" "%CEL(request.id)%" "%CEL(request.host)%" "%CEL(upstream.address)%"
```
Comment (zirain) [inline internal/xds/translator/testdata/in/xds-ir/accesslog-formatters.yaml:6]: can you add command like `%REQ_WITHOUT_QUERY` and `%METADATA`

**Comment by zirain** [comment]
this reminder should take a look if open telemetry access logger support this formatter.

File: internal/xds/translator/accesslog.go [hunk 3]
```diff
@@ -258,3 +355,11 @@ func processClusterForAccessLog(tCtx *types.ResourceVersionTable, al *ir.AccessL
...
 	return nil
 }
+
+// MessageToAny converts from proto message to proto Any
+// TODO: find a better way to handle error when translating access logs
+func MessageToAny(src proto.Message) *anypb.Any {
+	any, _ := anypb.New(src)
+
+	return any
+}
```
Comment (zirain) [inline internal/xds/translator/accesslog.go:365]: nit: use protocov.ToAny

**Comment by aoledk** [comment]
> this reminder should take a look if open telemetry access logger support this formatter.

I've checked Envoy docs about access log, it's not an option to set formatter for open telemetry access log [^1].

[^1]: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/open_telemetry/v3/logs_service.proto#extension-envoy-access-loggers-open-telemetry

**Comment by zirain** [comment]
> > this reminder should take a look if open telemetry access logger support this formatter.
> 
> I've checked Envoy docs about access log, it's not an option to set formatter for open telemetry access log [1](#user-content-fn-1-75d1026b992c03c5d5b15e61d9e78ccf).
> 
> ## Footnotes
> 1. https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/open_telemetry/v3/logs_service.proto#extension-envoy-access-loggers-open-telemetry [↩](#user-content-fnref-1-75d1026b992c03c5d5b15e61d9e78ccf)

but `CEL` `METADATA` should work on it.

**Comment by aoledk** [comment]
> but `CEL` `METADATA` should work on it.

I need to validate whether `CEL` or `METADATA` works with open telemetry access log under specific envoy configs.

I'll updated when it's done.

**Comment by zirain** [comment]
> > but `CEL` `METADATA` should work on it.
> 
> I need to validate whether `CEL` or `METADATA` works with open telemetry access log under specific envoy configs.
> 
> I'll updated when it's done.
> 

it's not blocker for this PR.

**Review by zirain — APPROVED** [review]
LGTM, thanks for adding this.

**Comment by aoledk** [comment]
When `CEL` or `METADATA` or `REQ_WITHOUT_QUERY` is configured in open telemetry access log, Envoy will panic [^1] with error `Not supported field in StreamInfo: CEL` or `Not supported field in StreamInfo: METADATA` or `Not supported field in StreamInfo: REQ_WITHOUT_QUERY`.

[^1]: https://github.com/envoyproxy/envoy/blob/2092d65bd4d476be8235ea541e5d25c096b513e6/source/common/formatter/stream_info_formatter.h#L277

**Comment by zirain** [comment]
> When `CEL` or `METADATA` or `REQ_WITHOUT_QUERY` is configured in open telemetry access log, Envoy will panic [1](#user-content-fn-1-4babb3164ff919be7c0088a073285e58) with error `Not supported field in StreamInfo: CEL` or `Not supported field in StreamInfo: METADATA` or `Not supported field in StreamInfo: REQ_WITHOUT_QUERY`.
> 
> ## Footnotes
> 1. https://github.com/envoyproxy/envoy/blob/2092d65bd4d476be8235ea541e5d25c096b513e6/source/common/formatter/stream_info_formatter.h#L277 [↩](#user-content-fnref-1-4babb3164ff919be7c0088a073285e58)

Thinks for trying this, envoy should support this.

**Review by shawnh2 — APPROVED** [review]
LGTM! Thanks for adding this!

**Comment by zirain** [comment]
this flaky seems just be fixed.

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Comment by arkodg** [comment]
would be great to add an example in our docs as a follow up

**Pull request merged**